### ISSUE TITLE
Surprised that $^N action works if string is global, fails if string is  local

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8170]
Migrated from [rt.perl.org#37543](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37543) (status was 'resolved')

Searchable as RT37543$

**Comment by p5pRT** [comment]
### From devanney@prodigy.net

#### Created by devanney@prodigy.net



Surprised that $^N does not work on local \(my\) variable in a sub but
instead 
requires a global variable\.  Following code works OK since '$foundstr'
is global\, but
remove '\#' \(making the $foundstr variable local to sub\) and code
crashes&#8203;:

\#START OF CODE
&findABC\('abcd'\);
&findABC\('abcd'\);  \# 2nd call should yield identical results

sub findABC
\{
&nbsp;  my $txt = shift;
&nbsp;  \# my \($foundstr\);  \# uncomment this statement to see problem

&nbsp;  if \($txt =~
&nbsp;    m/
&nbsp;    \(ABC\)  \(?\{$foundstr = $^N\}\)
&nbsp;    /ix
&nbsp;    \)
&nbsp;    \{
&nbsp;        printf "after regex \\$foundstr = '$foundstr'\\n";
&nbsp;    \};
&nbsp;  \}

\# END OF CODE

\# EXAMPLE OUTPUT&#8203;:
Case 1\) with global $foundstr&#8203;:

after regex $foundstr = 'abc'
after regex $foundstr = 'abc'

Case 2\) with "my $foundstr;" statement added to sub&#8203;:

after regex $foundstr = 'abc'
Use of uninitialized value in concatenation \(\.\) or string at problem\.pl
line 15\.

after regex $foundstr = ''

\# END OF EXAMPLE OUTPUT

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.7:

Configured by builder at Mon Jun  6 13:36:05 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=MSWin32, osvers=5.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32
-D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DBUILT_BY_ACTIVESTATE
-DNO_HASH_SEED -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf 
-libpath:"C:\Perl\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib 
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib 
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib 
version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug
-opt:ref,icf  -libpath:"C:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    #  if !defined(PERL_DARWIN)
    Iin_load_module moved for compatibility with build 806
    #  endif
    #  if defined(__hpux)
    Avoid signal flag SA_RESTART for older versions of HP-UX
    #  endif
    PerlEx hacks for CGI::Carp
    Less verbose ExtUtils::Install and Pod::Find
    instmodsh upgraded from ExtUtils-MakeMaker-6.25
    24699 ICMP_UNREACHABLE handling in Net::Ping
    21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.7:
    C:/perl/lib
    C:/perl/site/lib
    .


Environment for perl v5.8.7:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
   
PATH=C:\Perl\bin\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\PROGRA~1\COMMON~1\Odbc\FILEMA~1;H:\Program
Files\ATI\Control Panel;C:\Program Files\Common Files\Ulead
Systems\MPEG;H:\Program Files\Ulead\Ulead DVD MovieFactory 3
SE;c:\wld\bin
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From japhy@perlmonk.org

On Oct 26\, via RT and UNEXPECTED\_DATA\_AFTERdevanney@&#8203;prodigy\.netsaid&#8203;:

> Surprised that $^N does not work on local \(my\) variable in a sub but
> instead
> requires a global variable\.  Following code works OK since '$foundstr'
> is global\, but
> remove '\#' \(making the $foundstr variable local to sub\) and code
> crashes&#8203;:

It's not that $^N doesn't work\.  It's that \(?\{ $foundstr = $^N \}\) ends up 
being a closure around the lexical $foundstr\.  This is a known bug\.  The 
first time the regex is compiled\, the \(?\{ \.\.\. \}\) refers to a specific 
lexical\.  Subsequent calls to the regex will still be operating on that 
first lexical\.

>        printf "after regex \\$foundstr = '$foundstr'\\n";

Don't use printf\(\) unless you mean "print a formatted string"\.

\-\- 
Jeff "japhy" Pinyan        %  How can we ever be the sold short or
RPI Acacia Brother \#734    %  the cheated\, we who for every service
http&#8203;://www\.perlmonks\.org/  %  have long ago been overpaid?
http&#8203;://princeton\.pm\.org/   %    \-\- Meister Eckhart

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From devanney@prodigy.net

thanks very much\.\.\.

"japhy@&#8203;perlmonk\.org via RT" wrote&#8203;:
> 
> On Oct 26\, via RT and UNEXPECTED\_DATA\_AFTERdevanney@&#8203;prodigy\.netsaid&#8203;:
> 
> > Surprised that $^N does not work on local \(my\) variable in a sub but
> > instead
> > requires a global variable\.  Following code works OK since '$foundstr'
> > is global\, but
> > remove '\#' \(making the $foundstr variable local to sub\) and code
> > crashes&#8203;:
> 
> It's not that $^N doesn't work\.  It's that \(?\{ $foundstr = $^N \}\) ends up
> being a closure around the lexical $foundstr\.  This is a known bug\.  The
> first time the regex is compiled\, the \(?\{ \.\.\. \}\) refers to a specific
> lexical\.  Subsequent calls to the regex will still be operating on that
> first lexical\.
> 
> >        printf "after regex \\$foundstr = '$foundstr'\\n";
> 
> Don't use printf\(\) unless you mean "print a formatted string"\.
> 
> \-\-
> Jeff "japhy" Pinyan        %  How can we ever be the sold short or
> RPI Acacia Brother \#734    %  the cheated\, we who for every service
> http&#8203;://www\.perlmonks\.org/  %  have long ago been overpaid?
> http&#8203;://princeton\.pm\.org/   %    \-\- Meister Eckhart

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
